python/ray/data/impl/block_list.py

```diff
@@ -43,16 +62,58 @@ def split(self, split_size: int) -> List["BlockList"]:
         return output
 
     def divide(self, block_idx: int) -> ("BlockList", "BlockList"):
+        """Divide into two BlockLists by the given block index.
+
+        Args:
+            block_idx: The block index to divide at.
+        """
         self._check_if_cleared()
         return (BlockList(self._blocks[:block_idx],
                           self._metadata[:block_idx]),
                 BlockList(self._blocks[block_idx:],
                           self._metadata[block_idx:]))
 
-    def __len__(self):
+    def iter_blocks(self) -> Iterator[ObjectRef[Block]]:
+        """Iterate over the blocks of this block list.
+
+        This blocks on the execution of the tasks generating block outputs.
+        The length of this iterator is not known until execution.
+        """
         self._check_if_cleared()
-        return len(self._blocks)
+        outer = self
+
+        class Iter:
+            def __init__(self):
+                self._base_iter = outer.iter_blocks_with_metadata()
+
+            def __iter__(self):
+                return self
 
-    def __iter__(self):
+            def __next__(self):
+                ref, meta = next(self._base_iter)
+                assert isinstance(ref, ray.ObjectRef), (ref, meta)
+                return ref
+
+        return Iter()
+
+    def iter_blocks_with_metadata(
+            self) -> Iterator[Tuple[ObjectRef[Block], BlockMetadata]]:
+        """Iterate over the blocks along with their runtime metadata.
+
+        This blocks on the execution of the tasks generating block outputs.
+        The length of this iterator is not known until execution.
+        """
         self._check_if_cleared()
-        return iter(self._blocks)
+        return zip(self._blocks, self._metadata)
+
+    def initial_num_blocks(self) -> int:
```

Comment:
Sure, although `BlockList` doesn't have the concept of initial vs executed blocks either since `BlockList` is eager. But initial + executed does seem a bit more acceptable than the partition concept here, so this seems good to me.